tools/osbuilder/dockerfiles/QAT/run.sh

```diff
@@ -0,0 +1,170 @@
+#!/bin/bash
+#
+# Copyright (c) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+set -e
+set -u
+
+# NOTE: Some env variables are set in the Dockerfile - those that are
+# intended to be over-rideable.
+export QAT_SRC=~/src/QAT
+export ROOTFS_DIR=~/src/rootfs
+export GOPATH=~/src/go
+export PATH=${PATH}:/usr/local/go/bin:${GOPATH}/bin
+
+katarepo=github.com/kata-containers/kata-containers
+katarepopath=${GOPATH}/src/${katarepo}
+
+testsrepo=github.com/kata-containers/tests
+testsrepopath=${GOPATH}/src/${testsrepo}
+
+grab_kata_repos()
+{
+    # Check out all the repos we will use now, so we can try and ensure they use the specified branch
+    # Only check out the branch needed, and make it shallow and thus space/bandwidth efficient
+    # Use a green prompt with white text for easy viewing
+    bin/echo -e "\n\e[1;42mClone and checkout Kata repos\e[0m" 
+    git clone --single-branch --branch $KATA_REPO_VERSION --depth=1 https://${katarepo} ${katarepopath}
+    git clone --single-branch --branch $KATA_REPO_VERSION --depth=1 https://${testsrepo} ${testsrepopath}
+}
+
+configure_kernel()
+{
+    cp /input/qat.conf ${katarepopath}/tools/packaging/kernel/configs/fragments/common/qat.conf
+    # We need yq and go to grab kernel versions etc.
+    ${testsrepopath}/.ci/install_yq.sh
+    ${testsrepopath}/.ci/install_go.sh -p
+    cd ${katarepopath}
+    /bin/echo -e "\n\e[1;42mDownload and configure Kata kernel with CRYPTO support\e[0m"
+    ./tools/packaging/kernel/build-kernel.sh setup
+}
+
+build_kernel()
+{
+    cd ${katarepopath}
+    LINUX_VER=$(ls -d kata-linux-*)
+    sed -i 's/EXTRAVERSION =/EXTRAVERSION = .qat.container/' $LINUX_VER/Makefile
+    /bin/echo -e "\n\e[1;42mBuild Kata kernel with CRYPTO support\e[0m" 
+    ./tools/packaging/kernel/build-kernel.sh build
+}
+
+build_rootfs()
+{
+    # Due to an issue with debootstrap unmounting /proc when running in a
+    # --privileged container, change into /proc to keep it from being umounted. 
+    # This should only be done for Ubuntu and Debian based OS's. Other OS 
+    # distributions had issues if building the rootfs from /proc
+
+    if [ "${ROOTFS_OS}" == "debian" ] || [ "${ROOTFS_OS}" == "ubuntu" ]; then 
+        cd /proc
+    fi
+    /bin/echo -e "\n\e[1;42mDownload ${ROOTFS_OS} based rootfs\e[0m"
+    SECCOMP=no EXTRA_PKGS='kmod' ${katarepopath}/tools/osbuilder/rootfs-builder/rootfs.sh $ROOTFS_OS 
+}
+
+grab_qat_drivers()
+{
+    /bin/echo -e "\n\e[1;42mDownload and extract the drivers\e[0m" 
+    mkdir -p $QAT_SRC
+    cd $QAT_SRC
+    curl -L $QAT_DRIVER_URL | tar zx
```

Comment:
This is neat and concise, but we're running as `root` in docker I think? It would still be potentially safer to download as a non-priv user and then elevate privs as needed.

Reply:
It does run as root in the Docker image by default.  I checked through my QAT instructions and I don't think I actually need to be root for any of these parts except the depmod instruction and the QAT compilation instructions.  For me to make this work with a non privileged user I would need to redo the Dockerfile to add in a non privileged user and add in some sudo commands at the right spots as well as test everything to make sure it works.

Reply:
I came up with a way to accomplish this. I had to modify the `Dockerfile` so that there is a default user which I call `qatbuilder`. I think set its password to be blank and add that user to the sudoers file.  I then added sudo commands only where they are necessary.   I tested it with fedora, debian, and ubuntu as rootfs and it worked.